File: android/src/main/kotlin/com/ridehovr/hovr_native_platform_view/PhoneEntryPlatformView.kt
package com.ridehovr.hovr_native_platform_view

import android.content.Context
import android.view.View
import androidx.compose.runtime.mutableStateOf
import androidx.compose.ui.platform.ComposeView
import com.ridehovr.hovr_native_platform_view.ui.PhoneEntryScreen
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

class PhoneEntryPlatformView(
    context: Context,
    messenger: BinaryMessenger,
    viewId: Int,
) : PlatformView {

    private val channel = MethodChannel(
        messenger,
        PhoneEntryChannelConstants.channelName(viewId),
    )
    private val externalShowValidationError = mutableStateOf(false)
    private val composeView = ComposeView(context)

    init {
        ComposePlatformViewHelper.configure(composeView, context)
        channel.setMethodCallHandler(::handleDartCall)
        composeView.setContent {
            PhoneEntryScreen(
                externalShowValidationError = externalShowValidationError.value,
                onSubmit = ::sendSubmission,
            )
        }
    }

    override fun getView(): View = composeView

    override fun dispose() {
        channel.setMethodCallHandler(null)
    }

    private fun sendSubmission(phoneNumber: String, dialCode: String, countryIso: String) {
        channel.invokeMethod(
            PhoneEntryChannelConstants.ON_PHONE_SUBMITTED,
            mapOf(
                "phoneNumber" to phoneNumber,
                "dialCode" to dialCode,
                "countryIso" to countryIso,
            ),
        )
    }

    private fun handleDartCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != PhoneEntryChannelConstants.UPDATE_SUBMISSION_STATUS) {
            result.notImplemented()
            return
        }

        val args = call.arguments as? Map<*, *> ?: emptyMap<Any, Any>()
        val status = args["status"] as? String ?: "error"
        externalShowValidationError.value = status != "success"
        result.success(null)
    }
}

File: ios/Classes/PhoneEntryPlatformView.swift
import Flutter
import SwiftUI
import UIKit

final class PhoneEntryPlatformView: NSObject, FlutterPlatformView {
    private let viewModel = PhoneEntryViewModel()
    private let hostingController: UIHostingController<PhoneEntryUIView>
    private let channel: FlutterMethodChannel

    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        messenger: FlutterBinaryMessenger
    ) {
        channel = FlutterMethodChannel(
            name: PhoneEntryChannelConstants.channelName(viewId: viewId),
            binaryMessenger: messenger
        )
        hostingController = UIHostingController(
            rootView: PhoneEntryUIView(viewModel: viewModel)
        )
        super.init()

        viewModel.onSubmit = { [weak self] phoneNumber, dialCode, countryIso in
            self?.sendSubmission(
                phoneNumber: phoneNumber,
                dialCode: dialCode,
                countryIso: countryIso
            )
        }

        channel.setMethodCallHandler { [weak self] call, result in
            self?.handleDartCall(call, result: result)
        }
        hostingController.view.frame = frame
        hostingController.view.backgroundColor = .clear
    }

    func view() -> UIView {
        hostingController.view
    }

    func dispose() {
        channel.setMethodCallHandler(nil)
        viewModel.onSubmit = nil
    }

    private func sendSubmission(
        phoneNumber: String,
        dialCode: String,
        countryIso: String
    ) {
        let payload: [String: Any] = [
            "phoneNumber": phoneNumber,
            "dialCode": dialCode,
            "countryIso": countryIso,
        ]
        channel.invokeMethod(PhoneEntryChannelConstants.onPhoneSubmitted, arguments: payload)
    }

    private func handleDartCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard call.method == PhoneEntryChannelConstants.updateSubmissionStatus else {
            result(FlutterMethodNotImplemented)
            return
        }

        let args = call.arguments as? [String: Any] ?? [:]
        let status = args["status"] as? String ?? "error"
        let message = args["message"] as? String ?? ""
        let isSuccess = status == "success"

        DispatchQueue.main.async { [weak self] in
            self?.viewModel.applySubmissionStatus(isSuccess: isSuccess, message: message)
        }
        result(nil)
    }
}
